### PR TITLE
Implement Vector.Ceiling / Vector.Floor

### LINKED ISSUE #20509: API Proposal: Ceil, Floor for Vector<T>
This is a proposal of the ideas discussed in issue dotnet/runtime#20147.

There are a few SIMD operations on `System.Numerics.Vector<T>` missing that are required to implement vectorized versions of `log`, `exp`, `sin`, and `cos`.

There is a C implementation of these methods using AVX intrinsics [here](http://software-lisc.fbk.eu/avx_mathfun/), which I have translated into a C# implemention using `Vector<T>` [here](https://github.com/mjmckp/VectorMathFuns).   In the implementation, I have added fake versions of the missing intrinsics.

@mellinoe suggested creating two separate issues for the missing methods (the other issue is dotnet/corefx#16835), so the purpose of this issue is to propose the addition of the following methods:

[Updated proposal with comments below]
```C#
namespace System.Numerics
{
    public static class Vector
    {
        public static Vector<float> Floor(Vector<float> value);
        public static Vector<double> Floor(Vector<double> value);
        public static Vector<float> Ceiling(Vector<float> value);
        public static Vector<double> Ceiling(Vector<double> value);
    }
}
```

which map directly to the following AVX intrinsics:
```
_mm256_floor_ps
_mm256_floor_pd
_mm256_ceil_ps
_mm256_ceil_pd
```
respectively.

### PR DESCRIPTION
**Pull request by Gnbrkm41**
Fixes https://github.com/dotnet/runtime/issues/20509

Utilises `ROUNDPS`/`ROUNDPD` on x64, `FRINTP`/`FRINTM` on ARM64.
Example Code:
```cs
using System;
using System.Numerics;

public static class Program
{
    private static void Main()
    {
        Vector<float> vec = new Vector<float>(4.5f);
        vec = Vector.Ceiling(vec);
        Console.WriteLine(vec[0]);
    }
}
```

Assembly generated: (W10 Pro Insiders 19559 x64, Intel i7-8700)
```asm
*************** After end code gen, before unwindEmit()
G_M27646_IG01:        ; func=00, offs=000000H, size=001FH, bbWeight=1    PerfScore 7.00, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG

IN0009: 000000 push     rbp
IN000a: 000001 sub      rsp, 80
IN000b: 000005 vzeroupper
IN000c: 000008 lea      rbp, [rsp+50H]
IN000d: 00000D xor      rax, rax
IN000e: 00000F mov      qword ptr [V00 rbp-30H], rax
IN000f: 000013 mov      qword ptr [V00+0x8 rbp-28H], rax
IN0010: 000017 mov      qword ptr [V00+0x10 rbp-20H], rax
IN0011: 00001B mov      qword ptr [V00+0x18 rbp-18H], rax

G_M27646_IG02:        ; offs=00001FH, size=0029H, bbWeight=1    PerfScore 21.25, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, isz

IN0001: 00001F vbroadcastss ymm0, ymmword ptr[reloc @RWD00]
IN0002: 000028 vmovupd  ymmword ptr[V00 rbp-30H], ymm0
IN0003: 00002D vmovupd  ymm0, ymmword ptr[V00 rbp-30H]
IN0004: 000032 vroundps ymm0, ymm0, 10
IN0005: 000038 vmovupd  ymmword ptr[V00 rbp-30H], ymm0
IN0006: 00003D vmovss   xmm0, dword ptr [rbp-30H]
IN0007: 000042 call     System.Console:WriteLine(float)
IN0008: 000047 nop

G_M27646_IG03:        ; offs=000048H, size=0009H, bbWeight=1    PerfScore 3.00, epilog, nogc, extend

IN0012: 000048 vzeroupper
IN0013: 00004B lea      rsp, [rbp]
IN0014: 00004F pop      rbp
IN0015: 000050 ret
```

cc @tannergooding, @danmosemsft

### CODE REVIEW

**Comment by Gnbrkm41** [comment]
Currently, if `COMPlus_EnableSse41=0` the assertion in `lsraxarch.cpp` L1975 `compiler->getSIMDSupportLevel() >= SIMD_SSE4_Supported` fails. Need to make it not treat as intrinsic if we don't have SSE4.1...

**Comment by Gnbrkm41** [comment]
Any clues on what `##[error]Failed to build "CoreCLR component".` is about for *nix builds? 🙄

**Comment by EgorBo** [comment]
I wonder if it makes sense to implement new Vector`1 APIs using System.Runtime.Intrinsics API in C#. So Mono will get these methods accelerated for free 🙂 cc @tannergooding

**Comment by Gnbrkm41** [comment]
> I wonder if it makes sense to implement new Vector`1 APIs using System.Runtime.Intrinsics API in C#

That would be very, very nice (as I'm struggling to implement the JIT intrinsics part) but I don't know if there was any progression on it; https://github.com/dotnet/runtime/issues/952 have been implemented by https://github.com/dotnet/coreclr/pull/27401 so it should be doable at this point.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs
Comment (Gnbrkm41) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs]: oops, should be marked intrinsic 🙄

File: src/coreclr/src/jit/simdcodegenxarch.cpp
Comment (EgorBo) [inline src/coreclr/src/jit/simdcodegenxarch.cpp]: I guess now [these](https://github.com/dotnet/runtime/blob/a1bc3cc248bcd5b1edbf26cd8e4303b9e8baecdb/src/coreclr/src/jit/codegenxarch.cpp#L7276-L7286) magic numbers can be replaced with these named constants

**Comment by Gnbrkm41** [comment]
Changed to managed implementation w/ HWIntrinsics, and it looks quite promising:
```asm
*************** After end code gen, before unwindEmit()
G_M27646_IG01:        ; func=00, offs=000000H, size=0007H, bbWeight=1    PerfScore 1.25, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG

IN0005: 000000 sub      rsp, 40
IN0006: 000004 vzeroupper

G_M27646_IG02:        ; offs=000007H, size=0015H, bbWeight=1    PerfScore 12.25, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref

IN0001: 000007 vbroadcastss ymm0, ymmword ptr[reloc @RWD00]
IN0002: 000010 vroundps ymm0, ymm0, 10
IN0003: 000016 call     System.Console:WriteLine(float)
IN0004: 00001B nop

G_M27646_IG03:        ; offs=00001CH, size=0008H, bbWeight=1    PerfScore 2.25, epilog, nogc, extend

IN0007: 00001C vzeroupper
IN0008: 00001F add      rsp, 40
IN0009: 000023 ret
```

Here's what the JIT intrinsics version look on checked build:
```asm
*************** After end code gen, before unwindEmit()
G_M27646_IG01:        ; func=00, offs=000000H, size=0007H, bbWeight=1    PerfScore 1.25, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG

IN0005: 000000 sub      rsp, 40
IN0006: 000004 vzeroupper

G_M27646_IG02:        ; offs=000007H, size=0015H, bbWeight=1    PerfScore 12.25, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref

IN0001: 000007 vbroadcastss ymm0, ymmword ptr[reloc @RWD00]
IN0002: 000010 vroundps ymm0, ymm0, 10
IN0003: 000016 call     System.Console:WriteLine(float)
IN0004: 00001B nop

G_M27646_IG03:        ; offs=00001CH, size=0008H, bbWeight=1    PerfScore 2.25, epilog, nogc, extend

IN0007: 00001C vzeroupper
IN0008: 00001F add      rsp, 40
IN0009: 000023 ret
```

So it actually ends up generating identical code.

**Comment by Gnbrkm41** [comment]
One concern I have regarding managed implementation at the moment is that we can't quite utilise SIMD instructions this way, since most of the ARM intrinsics aren't exposed yet.

**Comment by tannergooding** [comment]
> I wonder if it makes sense to implement new Vector`1 APIs using System.Runtime.Intrinsics API in C# instead. So Mono will get these methods accelerated for free

I had previously attempted this for some methods in: https://github.com/dotnet/coreclr/pull/27483. @jkotas, @CarolEidt, and @AndyAyersMS had raised concerns about there being more complex trees and various getting missed because of this.
I know there have been a few improvements to inlining since then, but they would still need to give sign-off before I would feel comfortable giving the go ahead.

> One concern I have regarding managed implementation at the moment is that we can't quite utilise SIMD instructions this way on ARM machines, since most of the ARM intrinsics aren't exposed yet. 

The ARM intrinsics are still a WIP and we haven't started the general porting of existing x86 code paths to have ARM equivalents yet. It is fine, IMO, to just accelerate x86 right now provided we have a tracking issue to also accelerate ARM64 once support for the relevant instructions is added.

**Comment by Gnbrkm41** [comment]
If we decide to implement in the JIT, how should it be handled for machines without SSE4.1? I'm not really seeing a good way to vectorise this with a set of other SIMD instructions.

**Comment by EgorBo** [comment]
> If we decide to implement in the JIT, how should it be handled for machines without SSE4.1? I'm not really seeing a good way to vectorise this with a set of other SIMD instructions.

simply call `Math.Ceiling`/`Math.Floor`  for each component?

> I had previously attempted this for some methods in: [dotnet/coreclr#27483](https://github.com/dotnet/coreclr/pull/27483). @jkotas, @CarolEidt, and @AndyAyersMS had raised concerns about there being more complex trees and various getting missed because of this.

I meant only for new members (for a start).

**Comment by tannergooding** [comment]
> If we decide to implement in the JIT, how should it be handled for machines without SSE4.1?

It should likely just be done in software. I would not feel comfortable codifying the SSE2 fallback for vectorized `Ceil`/`Floor` in the JIT.

The scalar implementation of `Ceil`/`Floor` for x64 Windows is essentially:
* https://github.com/amd/win-libm/blob/master/ceilf.c
* https://github.com/amd/win-libm/blob/master/floorf.c

For pre-SSE4.1 hardware, you would need to essentially take this algorithm and vectorize it (which, at first glance, should just involve a couple of masks and or'ing the results of the success/failure code paths together).

That being said, I'm also not particularly concerned about that code path. SSE4.1 came out in ~2007 and so is nearly 13 years old. Even the x86 emulation layer in Windows 10 on ARM supports these instructions. Additionally, for S.P.Corelib with R2R and some AOT scenarios there is support for the SSE4.1 code path via a runtime check against a cached CPUID value; so the non-SSE4.1 codepath won't be hit in that scenario either. (Happy to hear input from others if there is some use-case/scenario I am missing).

**Comment by tannergooding** [comment]
> I meant only for new members (for a start).

I think its the same in either scenario. There are many benefits to having the implementations in managed code; including lowering the entry barrier, faster iteration, improved codegen in several scenarios, and easier sharing with other runtimes but there are some drawbacks with the current JIT that might make it unacceptable, even if only done for a few methods initially.

We'd need sign-off from the people I tagged above before we could start taking that route.

File: src/coreclr/src/jit/simd.cpp
Comment (tannergooding) [inline src/coreclr/src/jit/simd.cpp]: This should have a check for the current SIMD level and should return `nullptr` if it isn't at least SSE4.1.

Essentially just `if (getSIMDSupportLevel() >= SIMD_SSE4_Supported)`, as is done in other places in this file.

File: src/coreclr/src/jit/simdcodegenxarch.cpp
Comment (Gnbrkm41) [reply src/coreclr/src/jit/simdcodegenxarch.cpp]: Where do you think those macros should go to, in order for those constants to be shared across `simdcodegenxarch.cpp` and `codegenxarch.cpp`? As someone that hadn't done a lot of C++ I'm not sure if it can go to `codegen.h` or some place.

**Comment by Gnbrkm41** [comment]
Formatting Linux x64 is failing, but the artifacts generated is 0 bytes? 👀

**Comment by Gnbrkm41** [comment]
....I definitely remember Visual Studio complaining about the `#endif` missing something at the end, but looks like it shouldn't be there. 😳

**Comment by tannergooding** [comment]
@Gnbrkm41, could you please clarify if you are blocked by anything on this?

**Comment by Gnbrkm41** [comment]
I don't think so. It is fully implemented (At least I think), has tests and the tests seem to run fine. The new APIs also have XML comments on it as well.

The only thing that would be nice would be addressing https://github.com/dotnet/runtime/pull/31993#discussion_r376783406 (replacing magic number elsewhere with constant defined here) but that strictly isn't necessary.

**Review by tannergooding — APPROVED** [review]
LGTM

**Comment by tannergooding** [comment]
CC. @CarolEidt, @echesakovMSFT, @dotnet/jit-contrib

**Review by CarolEidt — APPROVED** [review]
LGTM - thanks!

**Comment by CarolEidt** [comment]
Regarding implementing the `Vector<T>` intrinsics in terms of the hardware intrinsics, I think that's something that we should tackle separately - not necessarily for new intrinsics, but just starting with a small set, and evaluating the behavior in terms of impact to the amount of IR generated and the code quality in the presence of inlining and wrapping, etc.

**Comment by tannergooding** [comment]
@Gnbrkm41, would you be able to rebase onto or merge this with dotnet/master. Its been a while since the last commit and I'd like to validate this against the latest bits before merging.

**Comment by Gnbrkm41** [comment]
@tannergooding - Done, now we wait 😄

File: src/coreclr/src/jit/simdcodegenxarch.cpp
Comment (echesakovMSFT) [reply src/coreclr/src/jit/simdcodegenxarch.cpp]: > 
> 
> I think probably instrsxarch.h would be a good place - other thoughts @dotnet/jit-contrib ?
> In any event, I think it's something that can be deferred if you'd prefer (and perhaps open an issue)

I would put the constants into emitxarch.h or codegen.h - they are supposed to be used together with a emitter or codegen function.

In my opinion, instrsxarch.h and other instrs*.h should be used for defining instructions' encodings only.

Comment (CarolEidt) [reply src/coreclr/src/jit/simdcodegenxarch.cpp]: @echesakovMSFT - I generally agree that instrs*.h should only be used for instruction encodings, but these are constants that are specific to the encodings. I don't think they belong in codegen.h, but emitxarch.h makes a lot of sense.

File: src/coreclr/src/jit/simdcodegenxarch.cpp
Comment (echesakovMSFT) [inline src/coreclr/src/jit/simdcodegenxarch.cpp]: *Nit:* This could be slightly simplified as
```c++
case SIMDIntrinsicCeil:
case SIMDIntrinsicFloor:
    if (compiler->getSIMDSupportLevel() >= SIMD_SSE4_Supported)
    {
        if (baseType == TYP_FLOAT)
        {
            result = INS_roundps;
        }
        else
        {
            assert(baseType == TYP_DOUBLE);
            result = INS_roundpd;
        }

        assert(ival != nullptr);
        *ival = (intrinsicId == SIMDIntrinsicCeil) ? ROUNDPS_TOWARD_POSITIVE_INFINITY_IMM : ROUNDPS_TOWARD_NEGATIVE_INFINITY_IMM;
    }
    break;

File: src/coreclr/src/jit/simdcodegenxarch.cpp
Comment (tannergooding) [inline src/coreclr/src/jit/simdcodegenxarch.cpp]:
```suggestion
                else
                {
                    assert(baseType == TYP_DOUBLE);
                    result = INS_roundpd;
                }
```

File: src/coreclr/src/jit/simdcodegenxarch.cpp
Comment (tannergooding) [reply src/coreclr/src/jit/simdcodegenxarch.cpp]: GitHub supports multiline suggestions now: https://twitter.com/natfriedman/status/1232727363626029056?s=20
See https://github.com/dotnet/runtime/pull/31993/files/a83bb51298fe2b63f6669a43f0b4983ee07015f9#r391776898, for an example

File: src/coreclr/src/jit/simdcodegenxarch.cpp
Comment (tannergooding) [reply src/coreclr/src/jit/simdcodegenxarch.cpp]: (as per https://github.com/dotnet/runtime/pull/31993/files/a83bb51298fe2b63f6669a43f0b4983ee07015f9#r391773461)

File: src/coreclr/src/jit/simdcodegenxarch.cpp
Comment (echesakovMSFT) [reply src/coreclr/src/jit/simdcodegenxarch.cpp]: @tannergooding Nice! Thanks

**Review by echesakov — APPROVED** [review]
LGTM

File: src/coreclr/src/jit/simdcodegenxarch.cpp
Comment (Gnbrkm41) [reply src/coreclr/src/jit/simdcodegenxarch.cpp]: I think I'll defer this for now. Not only that I don't fully understand the codebase yet, I cannot tell if including another header (if we were to move it to emitxarch) just for that constant is a good idea as I lack experience with C++.

File: src/coreclr/src/jit/simdcodegenxarch.cpp
Comment (Gnbrkm41) [reply src/coreclr/src/jit/simdcodegenxarch.cpp]: Applied suggestion. Thanks!

**Review by stephentoub — APPROVED** [review]

**Comment by Gnbrkm41** [comment]
@stephentoub, was there a need to re-run the tests? I am happy to rebase this and run tests again if that is better.

**Comment by stephentoub** [comment]
> was there a need to re-run the tests? 

They hadn't been run in four days.  I just wanted to make sure nothing crept in since that would cause merging this to break the build.

**Comment by Gnbrkm41** [comment]
With no changes made on this branch, I'm not sure if merely re-running tests will actually make any difference? (Not including any changes in the testing environments, of course) The branch is ~70 commits behind the upstream master branch.

I just rebased this branch on master. Hopefully that'll make things safer :^).

**Comment by stephentoub** [comment]
> With no changes made on this branch, I'm not sure if merely re-running tests will actually make any difference?

That's why I closed it and re-opened it (rather than retriggering the tests in DevOps).  CI will effectively rebase it on the latest master in that case.

**Comment by Gnbrkm41** [comment]
> CI will effectively rebase it on the latest master in that case.
Oooooh, I did not know that. That makes sense. Thanks!

**Comment by tannergooding** [comment]
Thanks for the contribution @Gnbrkm41!

**Pull request merged**
